text/0000-unsafe-fields.md

```diff
@@ -0,0 +1,137 @@
+- Feature Name: `unsafe_fields`
+- Start Date: 2023-07-13
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+Fields may be declared `unsafe`. Unsafe fields may only be mutated (excluding interior mutability)
+or initialized in an unsafe context. Reading the value of an unsafe field may occur in either safe
+or unsafe contexts. An unsafe field may be relied upon as a safety invariant in other unsafe code.
+
+# Motivation
+
+Emphasis on safety is a key strength of Rust. A major point here is that any code path that can
+result in undefined behavior must be explicitly marked with the `unsafe` keyword. However, the
+current system is insufficient. While Rust provides the `unsafe` keyword at the function level,
+there is currently no mechanism to mark fields as `unsafe`.
+
+For a real-world example, consider the `Vec` type in the standard library. It has a `len` field that
+is used to store the number of elements present. Setting this field is exposed publicly in the
+`Vec::set_len` method, which has safety requirements:
+
+- `new_len` must be less than or equal to `capacity()`.
+- The elements at `old_len..new_len` must be initialized.
+
+This field is safe to read, but unsafe to mutate or initialize due to the invariants. These
+invariants cannot be expressed in the type system, so they must be enforced manually. Failure to do
+so may result in undefined behavior elsewhere in `Vec`.
+
+By introducing unsafe fields, Rust can improve the situation where a field that is otherwise safe is
+used as a safety invariant.
+
+# Guide-level explanation
+
+Fields may be declared `unsafe`. Unsafe fields may only be initialized or accessed mutably in an
+unsafe context. Reading the value of an unsafe field may occur in either safe or unsafe contexts. An
+unsafe field may be relied upon as a safety invariant in other unsafe code.
+
+Here is an example to illustrate usage:
+
+```rust
+struct Foo {
+    safe_field: u32,
+    /// Safety: Value must be an odd number.
+    unsafe unsafe_field: u32,
+}
+
+// Unsafe field initialization requires an `unsafe` block.
+// Safety: `unsafe_field` is odd.
+let mut foo = unsafe {
```

Comment:
I’m not sure I like that the entire struct expression is now inside an unsafe block (though I’m not sure what a better syntax would be). If the safety invariant requires the entire struct, this makes sense. However, if it is more specific, a larger unsafe block also allows unsafe code usage to initialise the safe fields, which should get their own unsafe blocks. Though perhaps this could just be linted against, e.g. don’t use unsafe expressions in a struct initialiser without putting them inside nested unsafe blocks.

Reply:
The alternative is to do this:

```rust
let mut foo = Foo {
    safe_field: 0,
    unsafe_field: unsafe { 1 },
};
```

That feels worse to me.

Presumably the safety invariant being promised is an invariant within the struct, even if it is not strictly required that that be the case.

Reply:
I agree that just unsafe in one field initialiser expression is insufficient as it means something very different. I do like however that it clearly shows which field the unsafety I applies to.

It reminds me a bit of unsafe blocks in unsafe functions (https://github.com/rust-lang/rfcs/pull/2585). Ideally, the fact that struct init is unsafe would not allow unsafe field init expressions. I doubt that special-casing struct inits to not propagate outer unsafe blocks would be backwards compatible, so a new lint would be the only avenue in this direction.

Some new syntax like this
```rust
let mut foo = unsafe Foo {
    safe_field: 0,
    unsafe unsafe_field: 1,
};
```
would communicate intent better but looks quite unnatural to me.

Perhaps in the future there might be explicit safe blocks to reassert a safe context within an unsafe block, so that it could be written as follows:
```rust
let mut foo = unsafe {
    Foo {
        safe_field: safe { /* some more complex expr here */ },
        unsafe_field: safe { /* some more complex expr here */ },
    }
};
```
which could be encouraged with clippy lints. Though for now just moving safe non-trivial struct field initialisers into variables that are initialised outside the unsafe block.

Overall, I think going with the original syntax of
```rust
let mut foo = unsafe {
    Foo {
        safe_field: 2,
        unsafe_field: 1,
    }
};
```
looks like a good and intuitive solution, though I still think that a lint against using unsafe expressions inside the struct expression without a nested unsafe block would still help.

Reply:
Maybe this?

```rust
let mut foo = unsafe Foo {
    safe_field: 0,
    unsafe_field: 1,
};
```

`unsafe` would be followed immediately by a struct expression, and doing so would _only_ indicate that unsafe fields may be initialized. It would not introduce an unsafe context.

Reply:
I like it - the syntax is still close enough to the unsafe block syntax (except when initialising tuple structs) that it's relatively intuitive what the unsafety applies to, but coupled strongly to the struct name so that it also makes sense why no unsafe context is introduced.

Reply:
I think there's something nice about
```rust
let mut foo = unsafe Foo {
    safe_field: 0,
    unsafe_field: 1,
};
```
but to me that would go with `unsafe struct Foo`, like how `unsafe trait Bar` leads to `unsafe impl Bar`.

Spitballing: if per-field safety is really needed, then maybe
```rust
let mut foo = Foo {
    safe_field: 0,
    unsafe unsafe_field: 1,
};
```

though maybe the answer is the same as for `unsafe { unsafe_fn_call(complicated_expr) }`: if you don't want the expr in the block, use a `let`.

After all, there's always field shorthand available, so you can do
```rust
let safe_field = unsafe { complex to construct };
let unsafe_field = easy and safe to construct;
unsafe { Foo { safe_field, unsafe_field } }
```

Perhaps the more important factor would be the workflow for the errors the programmer gets when changing a (internal and thus not semver break) field to `unsafe`.  Anything at the *struct* level wouldn't give new "hey, you need `unsafe` here" if you already had another field that was unsafe.

Of course, if the safety is at the struct level (not the field level) then that doesn't come up.